src/test/memory/test_mmap.rs

```diff
@@ -0,0 +1,354 @@
+    let (temp_fd, path) = nix::unistd::mkstemp(template.as_ref()).map_err(|e| e.to_string())?;
+    nix::errno::Errno::result(temp_fd).map_err(|e| e.to_string())?;
+
+    /* Make sure there is enough space to write after the mmap. */
+    {
+        nix::fcntl::posix_fallocate(temp_fd, 0, buf.len() as i64).map_err(|e| e.to_string())?;
+    }
+
+    // Write buffer contents to file.
+    nix::unistd::write(temp_fd, &buf).map_err(|e| e.to_string())?;
+
+    // Seek to beginning of file.
+    let rv =
+        nix::unistd::lseek(temp_fd, 0, nix::unistd::Whence::SeekSet).map_err(|e| e.to_string())?;
+    nix::errno::Errno::result(rv).map_err(|e| e.to_string())?;
+
+    // Read contents from file.
+    let mut mapbuf = vec![0_u8; buf.len()];
+    let rv = nix::unistd::read(temp_fd, mapbuf.as_mut_slice()).map_err(|e| e.to_string())?;
+    nix::errno::Errno::result(rv).map_err(|e| e.to_string())?;
+
+    nix::unistd::unlink(&path).map_err(|e| e.to_string())?;
+    nix::unistd::close(temp_fd).map_err(|e| e.to_string())?;
+
+    Ok(())
+}
+
+fn test_mmap_file(offset: libc::off_t) -> Result<(), String> {
+    let template = b"test_mmapXXXXXX";
+
+    /* Get a file that we can mmap and write into. */
+    let (temp_fd, path) = nix::unistd::mkstemp(template.as_ref()).map_err(|e| e.to_string())?;
+    nix::errno::Errno::result(temp_fd).map_err(|e| e.to_string())?;
+
+    /* Make sure there is enough space to write after the mmap. */
+    {
+        nix::fcntl::posix_fallocate(temp_fd, offset, MAPLEN as i64).map_err(|e| e.to_string())?;
+    }
+
+    /* Init a msg to write. */
+    let msg = b"Hello new world!";
+
+    /* Do the mmap and write the message into the resulting mem location. */
+    let mapbuf = unsafe {
+        libc::mmap(
+            std::ptr::null_mut(),
+            MAPLEN,
+            libc::PROT_READ | libc::PROT_WRITE,
+            libc::MAP_SHARED,
+            temp_fd,
+            offset,
+        )
+    };
+
+    assert!(mapbuf != libc::MAP_FAILED);
+
+    let map = unsafe { std::slice::from_raw_parts_mut::<u8>(mapbuf as *mut u8, MAPLEN) };
+    map.copy_from_slice(msg.as_ref());
+
+    let rv = unsafe { libc::munmap(mapbuf, MAPLEN) };
+    nix::errno::Errno::result(rv).map_err(|e| e.to_string())?;
+
+    let rv = nix::unistd::lseek(temp_fd, offset, nix::unistd::Whence::SeekSet)
+        .map_err(|e| e.to_string())?;
+    nix::errno::Errno::result(rv).map_err(|e| e.to_string())?;
+
+    let mut rdbuf = [0_u8; MAPLEN];
+    let rv = nix::unistd::read(temp_fd, &mut rdbuf[..]).map_err(|e| e.to_string())?;
+    nix::errno::Errno::result(rv).map_err(|e| e.to_string())?;
+
+    assert_eq!(msg, &rdbuf);
+
+    nix::unistd::unlink(&path).map_err(|e| e.to_string())?;
+    nix::unistd::close(temp_fd).map_err(|e| e.to_string())?;
+
+    Ok(())
+}
+
+fn page_size() -> usize {
+    match nix::unistd::sysconf(nix::unistd::SysconfVar::PAGE_SIZE).unwrap() {
+        Some(p) => p as usize,
+        None => 0,
```

Comment:
Probably better to panic in this case; returning 0 will just result in more-difficult-to-debug crashes a little further along

Reply:
e.g. I think you can just drop the `match` and instead add a second `unwrap`